ENH:  einsum_sumprod.c.src to cpp in numpy.multiarray see #29528

athurdekoos opened the pull request:
This PR addresses a part of #29528.

### Description
This change ports einsum_sumprod from the generated einsum_sumprod.c.src file to a C++ template source file (einsum_sumprod.cpp).  The goal is to improve readability and maintainability while preserving the existing behavior and performance characteristics.

There are no intended changes to the public API. 

### Summary of changes:
- Replacement of `einsum_sumprod.c.src` with  `einsum_sumprod.cpp`.
- Original function structure was preserved as closely as possible.
  - Function patterns remain the same.
  - Function names have been preserved where possible.
  - Variable names have been preserved where possible.
- Logic was maintained as close to the original as possible. 
- Some larger functions have been refactored into smaller discrete implementations.
- Helper functions added. 
- If possible, macro based optimizations have been preserved. 
- Introduced new macro implementations primarily to handle NPY_SIMD_F32 and NPY_SIMD_F64.
- Optimizations that were only present in some code paths in the original implementation have been ported to the corresponding functions in the new structure when possible. 
- Expanded and clarified some comments.
- Some templated parameters from the original code were removed where they were not implement or did not materially improve performance or clarity.

### Notes: 
- Prioritization of this port was readability over aggressive abstraction.
- When possible templating is resolved at compile time. 

### Considerations not implemented: 
- Considered introducing C++ namespaces but did not do so to keep in line with the overall C-API style. 
  - Some function names as a result are a long but descriptive. 
- Considered using templates as a replacement for some macros, but opted against this where it hurt clarity

### Benchmarks 
- Across multiple machines were no major performance changes were observed.
- However, on my personal machine, benchmarks seem to improve dramatically in several cases with the exception of a small inconsistent regression 

 
| Change | Before [fabf1844] <main> | After [fad21058] <einsum_sumprod_to_cpp> | Ratio | Benchmark (Parameter)                      |
|--------|---------------------------|-------------------------------------------|--------|--------------------------------------------|
| +      | 194±7μs                  | 220±10μs                                  | 1.13   | bench_linalg.Eindot.time_inner_trans_a_a   |
| -      | 25.5±1μs                 | 23.5±0.3μs                                | 0.92   | bench_linalg.Eindot.time_dot_d_dot_b_c     |
| -      | 1.10±0.1ms               | 987±30μs                                  | 0.89   | bench_linalg.Eindot.time_dot_trans_at_a    |
| -      | 106±10μs                 | 70.8±7μs                                  | 0.66   | bench_linalg.Eindot.time_dot_trans_a_atc   |
| -      | 1.44±0.02ms              | 899±30μs                                  | 0.62   | bench_linalg.Eindot.time_einsum_i_ij_j     |
| -      | 110±5ms                  | 33.1±3ms                                  | 0.30   | bench_linalg.Eindot.time_einsum_ijk_jil_kl |
| -      | 114±6ms                  | 6.11±0.08ms                               | 0.05   | bench_linalg.Eindot.time_einsum_ij_jk_a_b  |


Please let me know if you'd like me to adjust naming or structure.

Commit:
ENH: einsum_sumprod.c.src to cpp in numpy.multiarray see #29528

Commit:
WIP: einsum_sumprod to cpp in Numpy.multiarray see #29528

Commit:
ENH: einsum_sumprod.c.src to cpp in numpy.multiarray see #29528

athurdekoos commented:
@inakleinbottle quick fyi ping to keep you apprised of my current status

inakleinbottle commented on numpy/_core/src/multiarray/einsum_sumprod.cpp:
```diff
@@ -0,0 +1,2281 @@
+ * and SIMDF64 */
+template <>
+struct SumSIMD<npy_float, npy_float> {
+    using SimdType = NpySIMDF32;
```
I think this, and a great deal of the other supporting structure will be useful elsewhere and should probably be moved to a header file that can just be included in all these replacement files as they are added. It might be worth having a discussion with the optimization team too for the SIMD stuff to avoid replicating work.

inakleinbottle commented on numpy/_core/src/multiarray/einsum_sumprod.cpp:
```diff
@@ -0,0 +1,2281 @@
+/* Template where (npy_double, npy_double) will allow the SIMD
+ * capable version.*/
+template <typename T, typename Temptype,
+          typename std::enable_if<std::is_same<T, Temptype>::value &&
```
The enable_if construct is a little odd here. First, since we have access to C++17, you might want to use the _t and _v variants which make this a bit more readable. However, here I think simple overloading would serve the same purpose. C++ always prefers concrete instantiations over templates if it is the best match, and enable_if constructions have a large cost for instantiation whereas overloading has a much smaller cost. I think though, that what is needed here is actually a partially specializable struct template that can be used as a customization point inside the `sum_of_arr` function. For instance:

```c++
template <typename T, typename TempType, typename SFINAE=void>
struct SumOfArr {
    static TempType eval(T* daa, npy_intp size) noexcept(?) {}
};
template <typename T>
static inline TempTypeOf<T> sum_of_arr(T* data, npy_intp count) noexcept (?) {
    using Helper = SumOfArray<T, TempTypeOf<T>>;
    return Helper::eval(data, count);
}
```
This is just an idea, but it would give you relatively low-overhead control the implementation based on the type and TempType (which I've shorthanded here as a trait, but could be left as a template argument if multiple choices are necessary).

inakleinbottle commented:
Thanks @athurdekoos for tagging me in. I had a quick look over the code and made a couple of what I hope are helpful comments, or at least things to think about. To be clear, I don't think these are necessary changes, so this isn't a review of any kind. Both comments are "looking ahead" to the other c.src modules too, so establishing some standard patterns and reusable components might be a good idea. Very happy discuss or explain further.

ngoldbaum commented on numpy/_core/src/multiarray/einsum_sumprod.cpp:
```diff
@@ -0,0 +1,2281 @@
+    return accum;
+}
+
+template <typename T, typename Temptype>
```
can we give these template variables more meaningful names? here and throughout this code

athurdekoos replied:
I agree this is a good idea.
 
I'm not sure what a better name would be though.  `@temptype@` is what the c.src file called them. 

My best idea would be something like `Accum_Type` since it looks like it's mostly used in conjunction with the `accum` or other accumulator linked intermediates. 

I assume this was done for greater precision but am not entirely sure

For reference here are the differences

| Type             | temptype       |
|------------------|----------------|
| npy_half         | npy_float      |
| npy_cfloat       | npy_float      |
| npy_cdouble      | npy_double     |
| npy_clongdouble  | npy_longdouble |

ngoldbaum reviewed (commented):
I skimmed over the entire thing but didn't go through in detail. I left a couple comments about naming.

Overall I think I prefer this even if it does end up substantially longer. The templates you started with are very hard to read and understand given the multiple levels of string substitution happening.

It might also help for future readability to add a top-level comment describing the structure of the templating and how various kinds of einsum operations are dispatched. An ASCII table/flowchart wouldn't hurt IMO.

Commit:
ENH: einsum_sumprod.c.src to cpp in numpy.multiarray see #29528

athurdekoos commented:
Changes made:
- Created new file `multiarray/simd_accum_traits.h` to host SIMD wrappers.  
  - Relocated SIMD related supporting structures to `multiarray/simd_accum_traits.h`.
- Removed all `enable_if`, replaced with templated structs. 
- Removed all `typename Temptype` replaced with `typename AccumType` for better identification. 
- Removed all mention of `scaller*` replaced with `elementwise*`

Apologies for large diff!

inakleinbottle commented on numpy/_core/src/multiarray/simd_accum_traits.h:
```diff
@@ -0,0 +1,116 @@
+/* Wrapper around the npyv float32 SIMD, provides a uniform interface between SIMDF32
+ * and SIMDF64 */
+template <>
+struct SumSIMD<npy_float, npy_float> {
```
This would require having a default implementation that doesn't do any SIMD ops too.

seiko2plus commented on numpy/_core/src/multiarray/einsum_sumprod.cpp:
```diff
@@ -0,0 +1,2181 @@
+#include <numpy/ndarraytypes.h> /* for NPY_NTYPES_LEGACY */
+#include <numpy/npy_common.h>
+
+#include "simd_accum_traits.h"
```
```suggestion
```
Therefore, this header should be removed.

seiko2plus commented on numpy/_core/src/multiarray/einsum_sumprod.cpp:
```diff
@@ -0,0 +1,2181 @@
+#include <array>
+#include <iostream>
+#include <numpy/arrayobject.h>
+#include <numpy/halffloat.h>
```
```suggestion
#include "common.hpp"
```
Use the C++ class `Half` instead, which is implemented by `common/half.hpp`, rather than the C wrapper.

seiko2plus commented on numpy/_core/src/multiarray/einsum_sumprod.cpp:
```diff
@@ -0,0 +1,2181 @@
+floating_point_sum_of_products_muladd(const typename SimdType::T *data,
+                                      typename SimdType::T *data_out,
+                                      typename SimdType::AccumType scalar,
+                                      npy_intp count)
```
```suggestion
template <typename Lane, typename LaneAcc>
static inline NPY_GCC_OPT_3 void
floating_point_sum_of_products_muladd(const Lane *data,
                                      Lane *data_out,
                                      LaneAcc scalar,
                                      SSize count)
```
Just deduce the scalar type directly; regarding `npy_intp`, you can use `SSize`, which maps to `Py_ssize_t` or `ptrdiff_t`.

seiko2plus commented on numpy/_core/src/multiarray/einsum_sumprod.cpp:
```diff
@@ -0,0 +1,2181 @@
+
+template <typename SimdType>
+static inline NPY_GCC_OPT_3 typename SimdType::T
+floating_point_sum_of_arr(const typename SimdType::T *data, npy_intp count)
```
```suggestion
template <typename Lane>
static inline NPY_GCC_OPT_3 Lane
floating_point_sum_of_arr(const Lane *data, ptrdiff_t count)
```
Just deduce the scalar type directly; regarding `npy_intp`, you can use `SSize`, which maps to `Py_ssize_t` or `ptrdiff_t`.

seiko2plus commented on numpy/_core/src/multiarray/einsum_sumprod.cpp:
```diff
@@ -0,0 +1,2181 @@
+from<npy_half, npy_float>(npy_half v)
+{
+    return npy_half_to_float(v);
+}
```
```suggestion
```
Use `static_cast` instead. The C++ class `np::Half` should be used directly rather than using the C wrapper. This is also important for inlining optimization.

seiko2plus reviewed (changes requested):
Thank you for working on it. A wrapper for C universal intrinsics cannot be accepted, as we agreed to use Google Highway as part of the C++ transition within NEP-54. I’ve left some suggestions below. I also see an opportunity to reduce the complexity of the C++ code by leveraging more C++17 features—for example, variadic templates.

athurdekoos committed:
ENH: Google highway SIMD functionality added to einsum_sumprod.cpp

-einsum_sumprod.cpp
  All SIMD fuctions in einsum_sumprod have been updated to google
highway, NEP-54 standard

-simd.ind.hpp
  New wrappers have been added to reproduce functionality

athurdekoos committed:
ENH: Google highway SIMD functionality added to einsum_sumprod.cpp

-einsum_sumprod.cpp
  added compile guards to highway path

athurdekoos committed:
ENH: Google highway SIMD functionality added to einsum_sumprod.cpp

-simd_accum_traits.h
 removed as it is no longer needed

athurdekoos committed:
ENH: Small format in simd.inc.hpp to force CI to run.

-simd.inc.hpp
 small formating update to force CI to run by adding a forward slash

mattip commented:
It would be nice to rerun the einsum benchmarks on the latest changes.

athurdekoos commented:
@mattip good idea!
I'm showing this result on my machine: `BENCHMARKS NOT SIGNIFICANTLY CHANGED`

mattip commented:
Hmm. How do we reconcile these two conflicting statements?
> I'm showing this result on my machine: BENCHMARKS NOT SIGNIFICANTLY CHANGED

> Given your benchmark, it seems game changer for my use cases
